src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations.cs

```diff
@@ -0,0 +1,187 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Analyzer.Utilities.PooledObjects;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.AnalyzerUtilities.FlowAnalysis.Analysis.InvocationCountAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.FlowAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeQuality.Analyzers.QualityGuidelines
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed partial class AvoidMultipleEnumerations : DiagnosticAnalyzer
+    {
+        // TODO: Find a good rule id.
+        private const string RuleId = "HAA1838";
+
+        // TODO: Polishing the text here.
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.AvoidMultipleEnumerationsTitle), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+        private static readonly LocalizableString s_description = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.AvoidMultipleEnumerationsMessage), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+
+        internal static readonly DiagnosticDescriptor MultipleEnumerableDescriptor = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_description,
+            DiagnosticCategory.Performance,
+            RuleLevel.Disabled,
+            description: null,
+            isPortedFxCopRule: false,
+            isDataflowRule: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray<DiagnosticDescriptor>.Empty.Add(MultipleEnumerableDescriptor);
+
+        private static readonly ImmutableArray<string> s_executedImmediateMethods = ImmutableArray<string>.Empty
```

Comment:
I would love an attribute that I can add to my own methods for this scenario.

`JetBrains.Annotations` has [`InstantHandleAttribute`](https://www.jetbrains.com/help/resharper/Reference__Code_Annotation_Attributes.html#InstantHandleAttribute) for this.

I don't suppose it'd be straightforward to add such an attribute to the framework, but we could support a well-known attribute name so that folks can define it in their own assemblies.

Reply:
The attribute approach sounds like a good way to provide supplement content.
For the 'well-known' attribute name way, I am not sure if this is a commonly used way.
I will leave this comment open until I get the answer.

Reply:
I believe it is common. The compiler uses this approach for nullability attributes, for example. Also for `IsExternalInitAttribute`. The question is whether such an attribute would be accepted into the framework. @stephentoub may have an opinion here.

Reply:
I'm not understanding the benefit of the attribute here.  Why is knowing whether a delegate will be invoked as part of the call helpful in general but also the gate by which this analyzer is deciding what to flag?

Reply:
Hi @stephentoub,
So this analyzer is targeting detect possible multiple Enumeration over IEnumerable instances.
And my understanding for Drew proposal is to have an attribute that could mark the IEnumerable argument, (for example(name could be different), [CauseEnumeration])
if we have a method like
```
void Foo()
{
      var baz = Enumerate.Range(1, 10);
     Bar(baz);
     Bar(baz);
}

void Bar([CauseEnumeration] IEnumerable<int> e)
{
// Do something cool
}
```
If there is such an attribute then the author could mark the paramter, and later this analyzer could use this to know the duplicate calls 'Bar(baz)' should trigger the warnings.
And for the benefits, I think
1. Some interprocedural analyses could be avoided if this attribute is seen.
2. If the analyzer couldn't find the source of the method symbol (from metadata), this could be used to improve the precision. 

Tag @drewnoakes if you have any supplements.

Reply:
What's an example of an API that takes an `IEnumerable<T>` and might not potentially enumerate it?  I'm still not understanding what benefit such a `[CauseEnumeration]` provides; if a method needs to take an `IEnumerable<T>` as an argument, I'd expect there to be at least the possibility it might enumerate it, or else it wouldn't need to take it as an argument.

Reply:
There are many well-known methods that cause Enumeration over IEnumerable<T>, like what is listed here in.
And the attribute just gives the user the ability to define their own 'well-known' Enumeration method. 

>  a method needs to take an IEnumerable<T> as an argument, I'd expect there to be at least the possibility it might enumerate it, or else it wouldn't need to take it as an argument.

A method could take IEnumerable<T> in and return IEnumerable<T> by just doing some LINQ query, like
```
IEnumerable<int> Bar(IEnumerable<string> input) => input.Select(i => i.Length); 
```

Reply:
> A method could take IEnumerable in and return IEnumerable by just doing some LINQ query

So you're trying to distinguish whether the enumeration happens as part of the call itself?  In your example, Bar does contribute to the enumeration of `input`, it just doesn't do it promptly: rather, enumerating the thing `Bar` produces then in turn enumerates `input`.

Reply:
> So you're trying to distinguish whether the enumeration happens as part of the call itself?

Yes

Reply:
What do you do then when you come to `SomeUnknownMethod(e).Any()`, where `SomeUnknownMethod` accepts and returns an `IEnumerable<T>`... without access to the code for `SomeUnknownMethod`, how do you know whether `Any()` ends up enumerating `e`?  This attribute wouldn't help with that.

I understand the goal.  It seems very niche to warrant adding such an attribute to the core libraries.

Reply:
If there is a truly unknown method, then the analyzer is likely to give the wrong answer.
1. Don't do anything when an `IEnumerable e` is passed to `SomeUnknownMethod(e)`. In such cases, we will have some false-negative predictions. The analyzer only brings benefits when the callee is 'Known'.
2. Always assuming `SomeUnknownMethod(e)` enumerates `e`. In such cases, we will have some false-positive predictions.
We have to set the diagnostic rule level to 'suggestion'
I think option 1 might be a good starting point.

Reply:
> We have to set the diagnostic rule level to 'suggestion'

_If_ this analyzer moves ahead, it wouldn't have a default higher than "suggestion", regardless of this.  In lots of code, it's perfectly reasonable to enumerate the same source multiple times; we can't know enough in the common case to say it's wrong.

Reply:
Make sense to me : )